File: src/endf/radionuclide_production.py
# SPDX-FileCopyrightText: 2023-2025 Paul Romano
# SPDX-License-Identifier: MIT

import warnings
from dataclasses import dataclass
from typing import Dict, Iterable, List, Optional, Tuple

from .fileutils import PathLike
from .function import Tabulated1D
from .material import Material

__all__ = ['RadionuclideProduction', 'radionuclide_production',
           'isomer_table', 'level_to_isomeric_state']


@dataclass
class RadionuclideProduction:
    """Production data for a single final state of one reaction.

    Joins the MF=8 identification of a radioactive product (its ZAP, LFS
    level number, and excitation energy) with the energy-dependent data
    evaluated for that state: the MF=9 yield multiplicity and/or the
    MF=10 production cross section. Note that LFS is a level index of
    the product nuclide, not an isomeric-state index; matching a level
    to a metastable state requires comparing :attr:`excitation_energy`
    against decay data.

    Attributes
    ----------
    ZAP : int
        1000*Z + A of the product nuclide
    LFS : int
        Level number of the final state (0 for the ground state)
    QM : float
        Mass-difference Q value in [eV]
    QI : float
        Reaction Q value for this state in [eV]
    ELFS : float or None
        Excitation energy of the final state in [eV] from MF=8, or None
        when the evaluation has no MF=8 subsection for this state
    yields : Tabulated1D or None
        MF=9 yield multiplicity of the reaction cross section as a
        function of incident energy in [eV]
    cross_section : Tabulated1D or None
        MF=10 production cross section in [b] as a function of incident
        energy in [eV]

    """

    ZAP: int
    LFS: int
    QM: float
    QI: float
    ELFS: Optional[float] = None
    yields: Optional[Tabulated1D] = None
    cross_section: Optional[Tabulated1D] = None

    @property
    def excitation_energy(self) -> float:
        """Excitation energy of the final state in [eV], taken from the
        MF=8 ELFS value when present and otherwise from QM minus QI."""
        if self.ELFS is not None:
            return self.ELFS
        return self.QM - self.QI


def radionuclide_production(
        material: Material) -> Dict[int, List[RadionuclideProduction]]:
    """Collect radionuclide production data from MF=8/9/10.

    For every reaction that has an MF=9 or MF=10 section, the final
    states are returned in the order they appear in the evaluation, with
    the MF=9 and MF=10 data for the same (ZAP, LFS) pair merged into one
    :class:`RadionuclideProduction` and the MF=8 excitation energy
    attached when available. The tabulated functions are returned
    exactly as evaluated.

    Parameters
    ----------
    material
        Material to read production data from

    Returns
    -------
    dict
        Mapping of MT numbers to lists of :class:`RadionuclideProduction`

    """
    by_mt: Dict[int, set] = {}
    for mf, mt in material.sections:
        if mf in (9, 10):
            by_mt.setdefault(mt, set()).add(mf)

    result = {}
    for mt in sorted(by_mt):
        # MF=8 links each (ZAP, LFS) pair to an excitation energy
        elfs = {}
        if (8, mt) in material.section_data:
            for subsection in material.section_data[8, mt]['subsections']:
                key = int(subsection['ZAP']), int(subsection['LFS'])
                elfs[key] = float(subsection['ELFS'])

        states: Dict[tuple, RadionuclideProduction] = {}
        ordered = []
        for mf in sorted(by_mt[mt]):
            for level in material.section_data[mf, mt]['levels']:
                key = int(level['IZAP']), int(level['LFS'])
                state = states.get(key)
                if state is None:
                    state = RadionuclideProduction(
                        ZAP=key[0],
                        LFS=key[1],
                        QM=level['QM'],
                        QI=level['QI'],
                        ELFS=elfs.get(key),
                    )
                    states[key] = state
                    ordered.append(state)
                if mf == 9:
                    state.yields = level['Y']
                else:
                    state.cross_section = level['sigma']
        result[mt] = ordered
    return result


def isomer_table(decay_files: Iterable[PathLike]) -> Dict[Tuple[int, int], dict]:
    """Build a table of isomeric states from decay data evaluations.

    MF=8 identifies a radioactive product by a *nuclear level* index (LFS), not
    by an isomeric-state ordinal, so relating production data to a named
    metastable nuclide needs the excitation energies of the isomers, which decay
    data provides.

    The absolute excitation energy of each isomer is recovered from its
    isomeric-transition decay mode (RTYP == 3) Q value, chained through that
    mode's final isomeric state (RFS) down to lower isomers. It is ``None`` for a
    pure-beta isomer, which has no isomeric-transition mode to measure.

    Parameters
    ----------
    decay_files
        Decay data evaluations. Only the metastable files are needed; ground
        states are implicit.

    Returns
    -------
    Mapping of ``(Z, A)`` to ``{LISO: {"LIS", "half_life", "E_iso"}}``, where
    LISO is the isomeric-state ordinal, LIS the nuclear level index, and E_iso
    the excitation energy in eV.

    """
    raw = {}
    for filename in decay_files:
        with warnings.catch_warnings():
            warnings.simplefilter("ignore")
            material = Material(str(filename))
        section = material.section_data.get((8, 457))
        if section is None:
            continue
        Z, A = divmod(int(section['ZA']), 1000)
        liso = int(section['LISO'])
        lis = int(section['LIS'])

        half_life = section.get('T1/2')
        if isinstance(half_life, (tuple, list)):
            half_life = float(half_life[0])
        elif half_life is not None:
            half_life = float(half_life)

        it_q = None
        it_rfs = 0
        for mode in section.get('modes', []) or []:
            if float(mode.get('RTYP', 0)) == 3.0:
                q = mode.get('Q')
                it_q = float(q[0] if isinstance(q, (tuple, list)) else q)
                it_rfs = int(float(mode.get('RFS', 0)))
                break

        raw.setdefault((Z, A), {})[liso] = {
            'LIS': lis, 'half_life': half_life, 'it_q': it_q, 'it_rfs': it_rfs,
        }

    table = {}
    for za, isomers in raw.items():
        resolved = {0: 0.0}
        out = {}
        for liso in sorted(isomers):
            info = isomers[liso]
            if liso == 0:
                energy = 0.0
            elif info['it_q'] is not None:
                energy = info['it_q'] + resolved.get(info['it_rfs'], 0.0)
            else:
                energy = None
            resolved[liso] = energy if energy is not None else 0.0
            out[liso] = {'LIS': info['LIS'], 'half_life': info['half_life'],
                         'E_iso': energy}
        table[za] = out
    return table


def level_to_isomeric_state(Z: int, A: int, lfs: int,
                            excitation_energy: Optional[float],
                            table: Dict[Tuple[int, int], dict], *,
                            tol_eV: float = 3000.0) -> int:
    """Map a production level to an isomeric-state ordinal (LISO).

    The ground state maps to 0. Otherwise the level's excitation energy is
    matched against the isomer energies in ``table``; failing that, the level
    index is compared against LIS; failing that, a nuclide with exactly one
    isomer maps to it. A level that resolves to none of these is treated as
    ground, on the basis that a short-lived level gamma-cascades down.

    Parameters
    ----------
    Z, A
        Atomic and mass number of the product
    lfs
        MF=8 level index of the product
    excitation_energy
        Excitation energy of the level in eV, or None if unknown
    table
        Isomer table from :func:`isomer_table`
    tol_eV
        Tolerance for the energy match

    Returns
    -------
    Isomeric-state ordinal, 0 for the ground state

    """
    isomers = table.get((Z, A))
    if not isomers:
        return 0
    metastable = {liso: d for liso, d in isomers.items() if liso > 0}
    if not metastable:
        return 0
    if lfs == 0 or excitation_energy is None or excitation_energy < 1000.0:
        return 0

    # 1. energy match against the decay isomer energies
    best = None
    for liso, d in metastable.items():
        if d['E_iso'] is not None:
            residual = abs(excitation_energy - d['E_iso'])
            if best is None or residual < best[1]:
                best = (liso, residual)
    if best is not None and best[1] <= tol_eV:
        return best[0]

    # 2. level index fallback
    for liso, d in metastable.items():
        if d['LIS'] == lfs:
            return liso

    # 3. single isomer fallback
    if len(metastable) == 1:
        return next(iter(metastable))

    # 4. unresolved, so cascade to ground
    return 0

File: crates/endf/src/radionuclide_production.rs
//! Radionuclide production: which nuclides a reaction leaves behind, in which
//! state, and how much of each.
//!
//! MF=8 identifies each radioactive product and MF=9 or MF=10 gives the
//! energy-dependent yield or production cross section. The three are joined
//! here, keyed by reaction, because a consumer wants them together.

use std::collections::{BTreeMap, BTreeSet};

use crate::error::Result;
use crate::function::Tabulated1D;
use crate::material::Material;

/// Production data for one final state of one reaction.
///
/// LFS is a level index of the *product* nuclide, not an isomeric-state
/// ordinal. Turning one into the other needs decay data — see
/// [`level_to_isomeric_state`].
#[derive(Debug, Clone, PartialEq, Default)]
pub struct RadionuclideProduction {
    /// `1000*Z + A` of the product nuclide.
    pub zap: i64,
    /// Level number of the final state; 0 is the ground state.
    pub lfs: i64,
    /// Mass-difference Q value in eV.
    pub qm: f64,
    /// Reaction Q value for this state, in eV.
    pub qi: f64,
    /// Excitation energy of the final state in eV, from MF=8. `None` when the
    /// evaluation has no MF=8 subsection for this state.
    pub elfs: Option<f64>,
    /// MF=9 yield, as a multiplier on the reaction cross section.
    pub yields: Option<Tabulated1D>,
    /// MF=10 production cross section in barns.
    pub cross_section: Option<Tabulated1D>,
}

impl RadionuclideProduction {
    /// Excitation energy of the final state in eV.
    ///
    /// The MF=8 value when the evaluation gave one, and `QM - QI` otherwise —
    /// the same quantity read off the Q values.
    pub fn excitation_energy(&self) -> f64 {
        self.elfs.unwrap_or(self.qm - self.qi)
    }
}

/// Collect the radionuclide production data of a material, by MT.
///
/// Every reaction with an MF=9 or MF=10 section is included, its final states
/// in the order the evaluation writes them. MF=9 and MF=10 data for the same
/// `(ZAP, LFS)` pair are merged into one entry, and the MF=8 excitation energy
/// attached where there is one. The tabulated functions are returned exactly
/// as evaluated.
pub fn radionuclide_production(material: &Material) -> BTreeMap<i32, Vec<RadionuclideProduction>> {
    let mut by_mt: BTreeMap<i32, BTreeSet<i32>> = BTreeMap::new();
    for &(mf, mt) in material.section_data.keys() {
        if mf == 9 || mf == 10 {
            by_mt.entry(mt).or_default().insert(mf);
        }
    }

    let mut result = BTreeMap::new();
    for (mt, files) in by_mt {
        // MF=8 links each (ZAP, LFS) pair to an excitation energy.
        let mut elfs: BTreeMap<(i64, i64), f64> = BTreeMap::new();
        if let Some(mf8) = material.mf8(mt) {
            for sub in &mf8.subsections {
                elfs.insert((sub.zap as i64, sub.lfs), sub.elfs);
            }
        }

        // Insertion order is the evaluation's order, which is what a reader
        // expects back, so the states are kept in a vector and located by a
        // side map rather than being sorted.
        let mut ordered: Vec<RadionuclideProduction> = Vec::new();
        let mut index: BTreeMap<(i64, i64), usize> = BTreeMap::new();
        for mf in files {
            let section = match mf {
                9 => material.mf9(mt),
                _ => material.mf10(mt),
            };
            let Some(section) = section else { continue };
            for level in &section.levels {
                let key = (level.izap, level.lfs);
                let i = *index.entry(key).or_insert_with(|| {
                    ordered.push(RadionuclideProduction {
                        zap: key.0,
                        lfs: key.1,
                        qm: level.qm,
                        qi: level.qi,
                        elfs: elfs.get(&key).copied(),
                        ..Default::default()
                    });
                    ordered.len() - 1
                });
                if mf == 9 {
                    ordered[i].yields = Some(level.func.clone());
                } else {
                    ordered[i].cross_section = Some(level.func.clone());
                }
            }
        }
        result.insert(mt, ordered);
    }
    result
}

/// One isomeric state of a nuclide, as decay data describes it.
#[derive(Debug, Clone, PartialEq, Default)]
pub struct Isomer {
    /// The nuclear level index the evaluation gives this state.
    pub lis: i64,
    /// Half-life in seconds, where the evaluation gave one.
    pub half_life: Option<f64>,
    /// Excitation energy in eV. `None` for a pure-beta isomer, which has no
    /// isomeric transition to measure it by.
    pub e_iso: Option<f64>,
}

/// Isomeric states by `(Z, A)`, then by isomeric-state ordinal (LISO).
pub type IsomerTable = BTreeMap<(i64, i64), BTreeMap<i64, Isomer>>;

/// What one decay file said, before the excitation energies are chained.
#[derive(Debug, Clone, Default)]
struct RawIsomer {
    lis: i64,
    half_life: Option<f64>,
    /// Q value of the isomeric-transition decay mode, when there is one.
    it_q: Option<f64>,
    /// The isomeric state that transition leaves behind.
    it_rfs: i64,
}

/// Build a table of isomeric states from decay data evaluations.
///
/// MF=8 identifies a radioactive product by a nuclear *level* index, not by an
/// isomeric-state ordinal, so relating production data to a named metastable
/// nuclide needs the excitation energies of the isomers — which is what decay
/// data provides.
///
/// Each isomer's absolute excitation energy is recovered from the Q value of
/// its isomeric-transition decay mode (RTYP = 3), chained through that mode's
/// final isomeric state down to lower isomers.
///
/// Only the metastable files are needed; ground states are implicit.
pub fn isomer_table<I, P>(decay_files: I) -> Result<IsomerTable>
where
    I: IntoIterator<Item = P>,
    P: AsRef<std::path::Path>,
{
    let mut raw: BTreeMap<(i64, i64), BTreeMap<i64, RawIsomer>> = BTreeMap::new();
    for filename in decay_files {
        let material = Material::from_file(filename.as_ref())?;
        let Some(section) = material.mf8_mt457() else {
            continue;
        };
        let (z, a) = (section.za / 1000, section.za % 1000);

        // The first isomeric transition is the one that fixes the energy; a
        // second would describe the same level.
        let it = section.modes.iter().find(|m| m.rtyp == 3.0);

        raw.entry((z, a)).or_default().insert(
            section.liso,
            RawIsomer {
                lis: section.lis,
                half_life: section.half_life.map(|(v, _)| v),
                it_q: it.map(|m| m.q.0),
                it_rfs: it.map_or(0, |m| m.rfs as i64),
            },
        );
    }

    let mut table = IsomerTable::new();
    for (za, isomers) in raw {
        // Energies resolved so far, which the chaining reads back.
        let mut resolved: BTreeMap<i64, f64> = BTreeMap::from([(0, 0.0)]);
        let mut out = BTreeMap::new();
        for (&liso, info) in &isomers {
            let energy = if liso == 0 {
                Some(0.0)
            } else {
                info.it_q
                    .map(|q| q + resolved.get(&info.it_rfs).copied().unwrap_or(0.0))
            };
            resolved.insert(liso, energy.unwrap_or(0.0));
            out.insert(
                liso,
                Isomer {
                    lis: info.lis,
                    half_life: info.half_life,
                    e_iso: energy,
                },
            );
        }
        table.insert(za, out);
    }
    Ok(table)
}

/// Default tolerance in eV for matching a level energy to an isomer energy.
pub const ISOMER_ENERGY_TOLERANCE: f64 = 3000.0;

/// Map a production level to an isomeric-state ordinal.
///
/// The ground state maps to 0. Otherwise the level's excitation energy is
/// matched against the isomer energies in `table`; failing that, the level
/// index is compared against LIS; failing that, a nuclide with exactly one
/// isomer maps to it. A level that resolves to none of these is treated as
/// ground, on the basis that a short-lived level gamma-cascades down.
pub fn level_to_isomeric_state(
    z: i64,
    a: i64,
    lfs: i64,
    excitation_energy: Option<f64>,
    table: &IsomerTable,
    tol_ev: f64,
) -> i64 {
    let Some(isomers) = table.get(&(z, a)) else {
        return 0;
    };
    let metastable: Vec<(&i64, &Isomer)> = isomers.iter().filter(|(&liso, _)| liso > 0).collect();
    if metastable.is_empty() {
        return 0;
    }
    // A level below a keV is not a metastable state; nor is the ground state,
    // whatever energy it is given.
    match excitation_energy {
        _ if lfs == 0 => return 0,
        None => return 0,
        Some(e) if e < 1000.0 => return 0,
        Some(_) => {}
    }
    let excitation_energy = excitation_energy.unwrap_or(0.0);

    // 1. The energy match against the decay isomer energies.
    let mut best: Option<(i64, f64)> = None;
    for (&liso, isomer) in &metastable {
        if let Some(e_iso) = isomer.e_iso {
            let residual = (excitation_energy - e_iso).abs();
            match best {
                Some((_, r)) if r <= residual => {}
                _ => best = Some((liso, residual)),
            }
        }
    }
    if let Some((liso, residual)) = best {
        if residual <= tol_ev {
            return liso;
        }
    }

    // 2. The level index.
    for (&liso, isomer) in &metastable {
        if isomer.lis == lfs {
            return liso;
        }
    }

    // 3. A nuclide with one isomer can only mean that one.
    if metastable.len() == 1 {
        return *metastable[0].0;
    }

    // 4. Unresolved, so cascade to ground.
    0
}

#[cfg(test)]
mod tests {
    use super::*;

    const IN115: &str = include_str!("../../../tests/n-049_In-115_trimmed.endf");

    #[test]
    fn joins_mf8_mf9_and_mf10_for_each_reaction() {
        let m = Material::from_str(IN115).unwrap();
        let production = radionuclide_production(&m);

        // The evaluation gives isomer production for three reactions, each to
        // a single excited state.
        assert_eq!(production.keys().copied().collect::<Vec<_>>(), [4, 16, 102]);
        assert!(production.values().all(|states| states.len() == 1));

        // Inelastic scattering and (n,2n) leave indium behind, and both are
        // given as MF=10 production cross sections.
        for (mt, zap) in [(4, 49115), (16, 49114)] {
            let state = &production[&mt][0];
            assert_eq!((state.zap, state.lfs), (zap, 1));
            assert!(state.cross_section.is_some());
            assert!(state.yields.is_none());
        }

        // Capture is given the other way, as an MF=9 yield on the MF=3 cross
        // section.
        let state = &production[&102][0];
        assert_eq!((state.zap, state.lfs), (49116, 1));
        assert!(state.yields.is_some());
        assert!(state.cross_section.is_none());

        // The excitation energy comes from MF=8's ELFS, not from QM - QI.
        assert_eq!(state.elfs, Some(127_269.7));
        assert_eq!(state.excitation_energy(), 127_269.7);
    }

    #[test]
    fn the_excitation_energy_falls_back_to_the_q_values() {
        // With no MF=8 subsection the energy is QM - QI, which is the same
        // quantity the evaluation would have written as ELFS.
        let state = RadionuclideProduction {
            qm: 6.0e6,
            qi: 5.8e6,
            elfs: None,
            ..Default::default()
        };
        assert_eq!(state.excitation_energy(), 2.0e5);

        let state = RadionuclideProduction {
            elfs: Some(1.0e5),
            ..state
        };
        assert_eq!(state.excitation_energy(), 1.0e5);
    }

    /// A two-isomer nuclide, as decay data would leave it: the second isomer
    /// transitions to the first, so its energy is the sum of the two Q values.
    fn two_isomers() -> IsomerTable {
        IsomerTable::from([(
            (95, 242),
            BTreeMap::from([
                (
                    1,
                    Isomer {
                        lis: 1,
                        half_life: Some(4.4e9),
                        e_iso: Some(48_600.0),
                    },
                ),
                (
                    2,
                    Isomer {
                        lis: 2,
                        half_life: Some(1.4e4),
                        e_iso: Some(2_200_000.0),
                    },
                ),
            ]),
        )])
    }

    #[test]
    fn an_energy_within_tolerance_picks_its_isomer() {
        let table = two_isomers();
        let at = |e: f64, lfs: i64| {
            level_to_isomeric_state(95, 242, lfs, Some(e), &table, ISOMER_ENERGY_TOLERANCE)
        };
        assert_eq!(at(48_600.0, 1), 1);
        assert_eq!(at(50_000.0, 1), 1);
        assert_eq!(at(2_200_500.0, 2), 2);
    }

    #[test]
    fn the_ground_state_and_low_levels_are_ground() {
        let table = two_isomers();
        let at = |e: Option<f64>, lfs: i64| {
            level_to_isomeric_state(95, 242, lfs, e, &table, ISOMER_ENERGY_TOLERANCE)
        };
        assert_eq!(at(Some(0.0), 0), 0);
        // A level index above zero but an energy too low to be metastable.
        assert_eq!(at(Some(500.0), 1), 0);
        // Nothing known about the energy.
        assert_eq!(at(None, 1), 0);
        // A nuclide the table has never heard of.
        assert_eq!(
            level_to_isomeric_state(1, 1, 1, Some(1.0e6), &table, ISOMER_ENERGY_TOLERANCE),
            0
        );
    }

    #[test]
    fn a_level_index_resolves_what_energy_cannot() {
        let table = two_isomers();
        // Far from either isomer energy, but the level index says which.
        assert_eq!(
            level_to_isomeric_state(95, 242, 2, Some(9.0e6), &table, ISOMER_ENERGY_TOLERANCE),
            2
        );
        // Neither energy nor index matches, and there are two isomers, so the
        // level is taken to cascade to ground.
        assert_eq!(
            level_to_isomeric_state(95, 242, 7, Some(9.0e6), &table, ISOMER_ENERGY_TOLERANCE),
            0
        );
    }

    #[test]
    fn a_nuclide_with_only_a_ground_state_is_ground() {
        let table = IsomerTable::from([(
            (26, 56),
            BTreeMap::from([(
                0,
                Isomer {
                    lis: 0,
                    half_life: None,
                    e_iso: Some(0.0),
                },
            )]),
        )]);
        assert_eq!(
            level_to_isomeric_state(26, 56, 3, Some(8.0e5), &table, ISOMER_ENERGY_TOLERANCE),
            0
        );
    }

    #[test]
    fn reads_the_isomers_of_in116_from_decay_data() {
        // The two metastable states of In116, which between them exercise both
        // branches: the first decays only by beta-, so it has no isomeric
        // transition to measure its energy by, and the second transitions down
        // to the first.
        let dir = std::path::Path::new(env!("CARGO_MANIFEST_DIR")).join("../../tests");
        let table = isomer_table([
            dir.join("dec-049_In_116m1.endf"),
            dir.join("dec-049_In_116m2.endf"),
        ])
        .unwrap();

        let in116 = &table[&(49, 116)];
        assert_eq!(in116.keys().copied().collect::<Vec<_>>(), [1, 2]);

        assert_eq!(in116[&1].lis, 1);
        assert_eq!(in116[&1].half_life, Some(3257.4));
        // Pure beta-, so the energy is unknown rather than zero.
        assert_eq!(in116[&1].e_iso, None);

        assert_eq!(in116[&2].lis, 4);
        assert_eq!(in116[&2].half_life, Some(2.18));
        // Its isomeric transition goes to state 1, whose energy is unknown and
        // so contributes nothing; the Q value stands alone.
        assert_eq!(in116[&2].e_iso, Some(162_393.0));
    }

    #[test]
    fn a_nuclide_with_one_isomer_resolves_to_it() {
        let table = IsomerTable::from([(
            (49, 116),
            BTreeMap::from([(
                1,
                Isomer {
                    lis: 1,
                    half_life: Some(3.3e3),
                    e_iso: None,
                },
            )]),
        )]);
        // No energy to match and no index to match, but only one candidate.
        assert_eq!(
            level_to_isomeric_state(49, 116, 4, Some(9.0e5), &table, ISOMER_ENERGY_TOLERANCE),
            1
        );
    }
}
